src/Client.php

```diff
@@ -12,57 +12,40 @@
 namespace JoliCode\Elastically;
 
 use Elastica\Client as ElasticaClient;
-use Elastica\Exception\RuntimeException;
-use Symfony\Component\PropertyInfo\Extractor\PhpDocExtractor;
-use Symfony\Component\Serializer\Encoder\JsonEncoder;
-use Symfony\Component\Serializer\Normalizer\ArrayDenormalizer;
-use Symfony\Component\Serializer\Normalizer\DateTimeNormalizer;
+use Psr\Log\LoggerInterface;
 use Symfony\Component\Serializer\Normalizer\DenormalizerInterface;
-use Symfony\Component\Serializer\Normalizer\ObjectNormalizer;
-use Symfony\Component\Serializer\Serializer;
 use Symfony\Component\Serializer\SerializerInterface;
 
 class Client extends ElasticaClient
 {
-    /* Elastically config keys */
-    public const CONFIG_MAPPINGS_DIRECTORY = 'elastically_mappings_directory';
-    public const CONFIG_INDEX_CLASS_MAPPING = 'elastically_index_class_mapping';
-    public const CONFIG_INDEX_PREFIX = 'elastically_index_prefix';
-    public const CONFIG_SERIALIZER_CONTEXT_PER_CLASS = 'elastically_serializer_context_per_class';
-    public const CONFIG_SERIALIZER = 'elastically_serializer';
-    public const CONFIG_BULK_SIZE = 'elastically_bulk_size';
-
-    private Indexer $indexer;
-    private IndexBuilder $indexBuilder;
+    /* Elastically config keys // BC Layer, to remove in 2.0 */
+    public const CONFIG_MAPPINGS_DIRECTORY = Factory::CONFIG_MAPPINGS_DIRECTORY;
+    public const CONFIG_SERIALIZER_CONTEXT_PER_CLASS = Factory::CONFIG_SERIALIZER_CONTEXT_PER_CLASS;
+    public const CONFIG_SERIALIZER = Factory::CONFIG_SERIALIZER;
+    public const CONFIG_BULK_SIZE = Factory::CONFIG_BULK_SIZE;
+    public const CONFIG_INDEX_PREFIX = Factory::CONFIG_INDEX_PREFIX;
+    public const CONFIG_INDEX_CLASS_MAPPING = Factory::CONFIG_INDEX_CLASS_MAPPING;
+
+    private Factory $factory;
     private ResultSetBuilder $resultSetBuilder;
-    private SerializerInterface $serializer;
-    private DenormalizerInterface $denormalizer;
+    private IndexNameMapper $indexNameMapper;
 
-    public function getIndexBuilder(): IndexBuilder
+    public function __construct($config = [], ?callable $callback = null, ?LoggerInterface $logger = null, ?ResultSetBuilder $resultSetBuilder = null, ?IndexNameMapper $indexNameMapper = null)
     {
-        if (!isset($this->indexBuilder)) {
-            $this->indexBuilder = new IndexBuilder($this, $this->getConfig(self::CONFIG_MAPPINGS_DIRECTORY));
-        }
+        parent::__construct($config, $callback, $logger);
 
-        return $this->indexBuilder;
-    }
-
-    public function getIndexer(): Indexer
-    {
-        if (!isset($this->indexer)) {
-            $this->indexer = new Indexer($this, $this->getSerializer(), $this->getConfigValue(self::CONFIG_BULK_SIZE, 100));
+        // BC Layer, to remove in 2.0
+        $config[Factory::CONFIG_CLIENT] = $this;
+        $this->factory = new Factory($config);
```

Comment:
Instead of using the `$config` array to push a Client instance, why not use a setter or a real property constructor on the Factory? To avoid injected the Client inside the Client.

Reply:
Good idea! ✅